highfidelity/hifi
# Multi-core Avatar Mixer

- This adds support for high scale avatar mixer by supporting multiple cores.
- This also changes the avatar mixer to a tight loop with sleep for scheduling instead of a timer based scheduling (more reliable timing)
- There are no changes to bandwidth allocation strategy.

Test Plan:
- Smoke test... must run tests with this sandbox (will work with recent master/dev-download builds) there are no protocol changes
- MUST test with multiple avatars/clients
- Can use the bots to test.
- Special tests:
    - Watch the avatar-mixer stats page in the domain > nodes... watch for *broadcast_loop_rate* remaining ~45.
    - as you load more avatars watch that it remains at 45 until the limits of the machine and then see that *broadcast_loop_rate* will begin to drop below 45 (on a 8 core machine this will be ~100 avatars)
    - In Domain > Settings > Advanced > Avatar Mixer, set the Number of Threads to a value other than 1 (not more than the number of cores on your machine), and verify that at that load the *broadcast_loop_rate* goes back to 45.

Make sure to test other avatar mixer features:
- verify that privacy bubble still works
- verify that when you change avatars people see the difference
- verify that hand controller joints move as expected
